### DISCUSSION TITLE
How to read database, page and blocks - examples?

### DISCUSSION THREAD

**Discussion by thysroes** [mariosimao/notion-sdk-php#156] [Q&A] [answered]
Hi there. I see the potential of this SDK and appreciate its development very much, so thank you. I have a hard time of getting it up and running unfortunately. The documentation looks very focused on adding and creating to Notion - not reading or querying.
I'm trying to use StatusFilter::equals("Published") but that seems the wrong syntax.
I've tried QueryAllPages but it gives me an error on one of the results in the database: "Fatal error: Uncaught ValueError: "0" is not a valid backing value for enum Notion\Users\UserType in /php-notion-sdk/vendor/mariosimao/notion-sdk-php/src/Users/User.php:42"

Connection with Notion works fine and print_r() of any of the results shows me the objects loaded correctly from Notion. I'm just having a really hard time showing the results because I the only example I found about reading and displaying _from_ Notion was this article: https://mario.engineering/notion-sdk-php/how-to/query-database.html 

More examples of how to filter Status="Published" and show database, page results or full block results from a page would really help me. Thanks!

**Comment by mariosimao** [comment, marked as answer]
Hi @thysroes, thanks for using the SDK!

## Status filter

Regarding the status filter, I will update it to make it easier to understand.

To address the `Status=Published` filter, you could try the following:

```php
<?php

use Notion\Notion;
use Notion\Databases\Query;
use Notion\Databases\Query\StatusFilter;

$token = $_ENV["NOTION_SECRET"];
$notion = Notion::create($token);

$databaseId = "c986d7b0-7051-4f18-b165-cc0b9503ffc2";
$database = $notion->databases()->find($databaseId);

$query = Query::create()->changeFilter(
    StatusFilter::property("Status")->equals("Published"),
);

$result = $notion->databases()->query($database, $query);

$pages = $result->pages; // filtered pages
```

## QueryAllPages error

About the error while querying all pages, this is probably an SDK error. Could you help me with a little bit of information so I can debug and solve this issue?

- What SDK and PHP are you using?
- What is the complete error stack trace?

## Displaying from Notion

I am unsure if I understood what you mean by "Displaying *from* Notion". With the SDK and the Notion API, you can read and write data from/to Notion. Getting to display the content from Notion in another place is out of the scope of the API.

**Reply by thysroes** [reply]
Hi Mario. An update on my more overarching question because I hope it helps you to get a sense of who is trying your SDK out: what I'm trying to do is trying to show a very simple page I made in Notion. I'm trying to use Notion as a CMS for my blog. So I create a page in Notion, and now would like to show its contents. Nothing more really! 

Some examples would reaaally help! But I see your documentation is very focused on "Creating", "Adding", and "Updating" TO Notion - but when it comes to just.... well... "Reading", "Showing", "Getting" content FROM Notion, I have found only one example at the bottom of the chapter "Find a page" which says: "Get page content" [#](https://mario.engineering/notion-sdk-php/how-to/find-a-page.html#get-page-content)

```
In Notion, pages behave like blocks where the content is their children. You can fetch page content with
$notion->blocks()->findChildren($pageId);
```

This works indeed and I'm getting an array back. So... now?

If you had an example available somewhere of how this would usually be done, it would clear so much up for me. Now I'm taking it baby step by baby step with chatGPT on the side, because I'm flying in the dark right now on what the regular idea should be right now.

Should I check for block type and then markup the contents manually? Is there a simple way to just show it raw? What's the right syntax? How to skip unsupported blocks? Etc, etc.

An example of how someone would just show a single page with a header, a paragraph with some links, and an image - all read FROM a Notion page with a certain pageId: it would be incredibly helpful to have.

Also - my very first page that was trying to show instantly gave me this error. It took me a while to figure out it was only doing this on one page with pageId (4f3a70c9-3274-47f6-9636-3d974ccb768a), I get the error:
 
```
Fatal error
: Uncaught Notion\Exceptions\BlockException: Invalid type: 'unsupported' in /vendor/mariosimao/notion-sdk-php/src/Exceptions/BlockException.php:22 Stack trace: #0 /vendor/mariosimao/notion-sdk-php/src/Blocks/BlockFactory.php(43): Notion\Exceptions\BlockException::invalidType() #1 /vendor/mariosimao/notion-sdk-php/src/Blocks/Client.php(48): Notion\Blocks\BlockFactory::fromArray() #2 [internal function]: Notion\Blocks\Client->Notion\Blocks\{closure}() #3 /vendor/mariosimao/notion-sdk-php/src/Blocks/Client.php(47): array_map() #4 /parseblogs.php(176): Notion\Blocks\Client->findChildren() #5 {main} thrown in
/vendor/mariosimao/notion-sdk-php/src/Exceptions/BlockException.php
on line
```

I understand not all types of Blocks are supported. But unsupported blocks throwing a fatal error is also not great, I'd rather just have them skip over them if a block is unsupported and leave the block for when its supported in the future. Now I'm not sure if I can even use the SDK for the purpose I'd like to use it for (displaying simple text/links/images blog posts built in Notion).

Any timeline on supporting images/video? And any way for me to figure out why that one page gives me this error?

Thanks again for your help and hopefully my feedback helps you too on this potentially very useful SDK.

**Reply by mariosimao** [reply]
Hi @thysroes, thanks for the feedback again. Now I understand better what you are trying to do.

## Unsupported blocks

It is really frustrating having a page that is not fully supported by the SDK. I created an issue for it (#157) and started developing this feature. I am currently on vacation and this should be released on the first week of March.

About Image and Video support, they are already supported by the SDK. If are facing issues with it, let me know.

## Reading blocks

I agree that the documentation may be missing content about this part. I will provide more examples on March, when I will have access back to my computer.

For your CMS usecase, something like the following lines of code may solve. You can elaborate on it to make it *cleaner*. There may be some PHP syntax errors since I am writing this on my phone.

```php
$content = $notion->blocks()->findChildren($pageId);

foreach ($content as $block) {
    switch ($block::class) {
        case \Notion\Blocks\Heading1::class:
            echo "<h1>" . $block->toString() . "</h1>";
            break;
        case \Notion\Blocks\Paragraph::class:
            echo "<p>" . $block->toString() . "</p>";
            break;
        case \Notion\Blocks\Image::class:
            echo '<img src="' . $block->file->url . '">';
            break;
    }
}
```

**Reply by mariosimao** [reply]
@thysroes, the support for unknown blocks and properties has been released on [version 1.5.0](https://github.com/mariosimao/notion-sdk-php/releases/tag/v1.5.0). With this new version, the errors should stop when fetching a page.

**Reply by thysroes** [reply]
You're the best. I'm going to try it out! 😀

(also, noticed that some earlier errors might have been produced by certain blogs not being shared to web correctly)

**Comment by thysroes** [comment]
Thank you, that works! 

ChatGPT helped me get to the PlainText title like:
$page->properties['Name']->title[0]->plainText

**Reply by mariosimao** [reply]
It blows my mind to see how ChatGPT can answer such a specific topic.

Are you using an IDE with autocomplete? The SDK is fully typed and the usage of suggestions/IntelliSense is highly recommended.

**Reply by thysroes** [reply]
And to answer your question:
Yes I'm using VS Code but not very in touch of everything it is capable of. I have to figure out how PHP Intellisense knows about your SDK and can autocomplete. 

I'm using SDK version 1.4.0 and PHP 8.2.2 by the way!

**Reply by mariosimao** [reply]
For writing and updating, the autocomplete should already be working if you have the extension installed without extra setup.

For reading, it is a bit more complicated because the page content is an array of mulitples types of blocks. One way for the IDE to know may be like this:

```php
$blocks = $notion->blocks()->findChildren($pageId);

foreach ($blocks as $block) {
    if ($block instaceof \Notion\Blocks\Paragraph) {
        // if you type '$block->' VS Code will suggest properties and methods from Paragraph blocks
    }
}
```

**Reply by thysroes** [reply]
Thank you Mario really appreciate your responses and your development work! And no rush my man, I'm traveling myself and doing this project outside of my regular work hours. But will give your suggestions a go this week and look forward to those examples in the documentation :).
